llvm/lib/Analysis/ConstantFolding.cpp

```diff
@@ -2209,6 +2209,11 @@ static Constant *ConstantFoldScalarCall1(StringRef Name,
     switch (IntrinsicID) {
       default: break;
       case Intrinsic::log:
+#if defined(__FLOAT128__) && defined(HAS_LOGF128)
+        if (Ty->isFP128Ty()) {
+          return ConstantFP::get(Ty, logf128(APF.convertToQuad()));
+        }
+#endif
         return ConstantFoldFP(log, APF, Ty);
```

Comment:
This falls through to using the double-precision log() for long double values on hosts which don't have float128; that seems wrong.

Reply:
Is this not already what was happening previously? Inside `ConstantFoldFP` the result is handled as a double. I'm not sure if you're asking me to start preserving the long double type in this pre-existing code or whether what I've introduced has changed established behaviour

Reply:
Without this patch, we wouldn't try to fold fp128 at all.  With this patch, we try to fold it using "double" routines.  (Previously, we would only fold 64-bit or 32-bit floats.)

Reply:
Ah I see what you mean now, thanks for clarifying. I've rearranged this to stop any fall-through happening